internal/cli/serverless/export/download_without_prompt.go

```diff
@@ -0,0 +1,248 @@
+	// The size of the batch to request download url
+	clusterID string
+	exportID  string
+	count     int
+	h         *internal.Helper
+
+	jobs    chan *downloadJob
+	results chan *downloadResult
+
+	// fileJobs is a list of jobs that will be set to the jobs channel
+	fileJobs  []*downloadJob
+	batchSize int
+}
+
+func NewDownloadPool(h *internal.Helper, path string,
+	concurrency int, exportID, clusterID string, count int, client cloud.TiDBCloudClient) (*downloadPool, error) {
+	if count <= 0 {
+		return nil, errors.New("no files to download")
+	}
+	if concurrency <= 0 {
+		concurrency = DefaultConcurrency
+	}
+
+	jobBufferSize := 2 * concurrency
+	if jobBufferSize > count {
+		jobBufferSize = count
+	}
+	jobs := make(chan *downloadJob, jobBufferSize)
+	results := make(chan *downloadResult, count)
+
+	return &downloadPool{
+		h:           h,
+		path:        path,
+		concurrency: concurrency,
+		client:      client,
+		clusterID:   clusterID,
+		exportID:    exportID,
+		jobs:        jobs,
+		results:     results,
+		batchSize:   BatchSize,
+		count:       count,
+	}, nil
+}
+
+type downloadJob struct {
+	fileName    string
+	downloadUrl *string
+	size        int
+	err         error
+}
+
+type downloadResult struct {
+	name   string
+	err    error
+	status JobStatus
+}
+
+func (r *downloadResult) GetResult() string {
+	if r.status == Succeeded {
+		return fmt.Sprintf("%s success", r.name)
+	}
+	if r.err == nil {
+		return fmt.Sprintf("%s %s", r.name, r.status)
+	}
+	return fmt.Sprintf("%s %s: %s", r.name, r.status, r.err.Error())
+}
+
+func (d *downloadPool) Start() error {
+	// create the path if not exist
+	err := util.CreateFolder(d.path)
+	if err != nil {
+		return err
+	}
+	pathName := d.path
+	if pathName == "" {
+		pathName = "current folder"
+	}
+	fmt.Fprintf(d.h.IOStreams.Out, color.GreenString("start to download files to %s:\n", pathName))
+	// start produce
+	go d.produce()
+	// start consumers:
+	for i := 0; i < d.concurrency; i++ {
+		wg.Add(1)
+		go d.consume()
+	}
+	// wait for all consumers to finish
+	wg.Wait()
+	close(d.results)
+	// summarize the download results
+	succeededCount := 0
+	failedCount := 0
+	skippedCount := 0
+	downloadResults := make([]*downloadResult, 0)
+	for result := range d.results {
+		switch result.status {
+		case Succeeded:
+			succeededCount++
+		case Failed:
+			failedCount++
+		case Skipped:
+			skippedCount++
+		}
+		downloadResults = append(downloadResults, result)
+	}
+	fmt.Fprintf(d.h.IOStreams.Out, generateDownloadSummary(succeededCount, skippedCount, failedCount))
+	index := 0
+	for _, f := range downloadResults {
+		if f.status != Succeeded {
+			index++
+			fmt.Fprintf(d.h.IOStreams.Out, "%d.%s\n", index, f.GetResult())
+		}
+	}
+	if failedCount > 0 {
+		return fmt.Errorf("%d file(s) failed to download", failedCount)
+	}
+	return nil
+}
+
+func (d *downloadPool) produce() {
+	pageSize := int32(d.batchSize)
+	var pageToken *string
+	ctx := context.Background()
+	for i := 0; i < d.count; i++ {
```

Comment:
I can understand this. But why not just keep listing and directly inserting the results into the channel?